File: src/fanout.rs
use std::sync::Arc;

use dashmap::DashMap;
use serde::Serialize;
use tracing::warn;

use crate::delta::DeltaOp;
use crate::frame::OutboundMsg;
use crate::write_pool::WriteHandle;

/// Default capacity for the per-session bounded channel.
/// When the channel is full, the subscriber is considered slow and evicted.
pub const DEFAULT_BACKPRESSURE_LIMIT: usize = 1024;

/// One active WATCH subscription: a write handle back to the owning session.
#[derive(Clone)]
pub struct Subscriber {
    pub session_id: String,
    pub stream_id: u32,
    pub write: Arc<WriteHandle>,
}

/// Registry of all active subscriptions, keyed by state name.
///
/// Designed for high-concurrency fanout:
/// - reads (broadcast) hold per-shard read locks only
/// - writes (subscribe/unsubscribe) hold per-shard write locks briefly
pub struct FanoutBus {
    subscriptions: DashMap<String, Vec<Subscriber>>,
}

/// Encode a DELTA body (version + ops) as MessagePack bytes.
///
/// The result is the shared body for all subscriber DELTA frames.  Call this
/// once per mutation, then pass the bytes to `broadcast_encoded` and/or
/// `frame::delta_bytes` for direct same-session delivery.
pub fn encode_delta_body(version: u64, ops: &[DeltaOp]) -> Vec<u8> {
    #[derive(Serialize)]
    struct DeltaBody<'a> {
        version: u64,
        ops: &'a [DeltaOp],
    }
    rmp_serde::to_vec_named(&DeltaBody { version, ops }).unwrap_or_default()
}

impl FanoutBus {
    pub fn new() -> Arc<Self> {
        Arc::new(FanoutBus {
            subscriptions: DashMap::new(),
        })
    }

    pub fn subscribe(&self, state_key: String, sub: Subscriber) {
        self.subscriptions.entry(state_key).or_default().push(sub);
    }

    pub fn unsubscribe(&self, state_key: &str, session_id: &str) {
        if let Some(mut subs) = self.subscriptions.get_mut(state_key) {
            subs.retain(|s| s.session_id != session_id);
        }
    }

    /// Remove every subscription belonging to a session (called on disconnect).
    pub fn remove_session(&self, session_id: &str) {
        for mut entry in self.subscriptions.iter_mut() {
            entry.value_mut().retain(|s| s.session_id != session_id);
        }
    }

    /// Broadcast a pre-encoded delta body to subscribers of `state_key`.
    ///
    /// Optimised for low P99:
    ///  1. Snapshot subscriber handles while holding the shard lock for the
    ///     shortest possible window (just pointer copies, no allocation).
    ///  2. Body is already encoded — per-subscriber cost is one Arc clone.
    ///
    /// `exclude_session`: if `Some(id)`, skip the subscriber whose
    /// `session_id == id`.  Use this when the triggering session will receive
    /// its own DELTA via a direct write.
    ///
    /// Slow subscribers (full channel) are cancelled and removed inline.
    pub fn broadcast_encoded(
        &self,
        state_key: &str,
        body_mp: &[u8],
        exclude_session: Option<&str>,
    ) {
        let snapshot: Vec<(u32, String, Arc<WriteHandle>)> = match self.subscriptions.get(state_key)
        {
            Some(subs) => subs
                .iter()
                .filter(|s| exclude_session.is_none_or(|ex| s.session_id != ex))
                .map(|s| (s.stream_id, s.session_id.clone(), Arc::clone(&s.write)))
                .collect(),
            None => return,
        };
        // DashMap shard lock released here.

        if snapshot.is_empty() {
            return;
        }

        // Wrap body_mp in Arc once — all subscribers share the bytes.
        let shared: Arc<[u8]> = Arc::from(body_mp);

        let mut slow_sessions: Vec<String> = Vec::new();

        for (stream_id, session_id, write) in snapshot {
            if !write.send(OutboundMsg::ArcDelta {
                stream_id,
                body_mp: Arc::clone(&shared),
            }) {
                warn!("Slow consumer {session_id} — cancelling session");
                slow_sessions.push(session_id);
            }
        }

        // Remove slow subscribers from this key's subscription list.
        if !slow_sessions.is_empty()
            && let Some(mut subs) = self.subscriptions.get_mut(state_key)
        {
            subs.retain(|s| !slow_sessions.contains(&s.session_id));
        }
    }

    /// Broadcast delta ops to every subscriber of `state_key`.
    ///
    /// Encodes the body internally; use `encode_delta_body` + `broadcast_encoded`
    /// directly when you also need the bytes for a same-session direct write.
    pub fn broadcast(&self, state_key: &str, version: u64, ops: &[DeltaOp]) {
        if ops.is_empty() {
            return;
        }
        let body_mp = encode_delta_body(version, ops);
        if body_mp.is_empty() {
            return;
        }
        self.broadcast_encoded(state_key, &body_mp, None);
    }

    pub fn subscriber_count(&self, state_key: &str) -> usize {
        self.subscriptions
            .get(state_key)
            .map(|s| s.len())
            .unwrap_or(0)
    }
}

File: sodp-go/fanout.go
package sodp

import (
	"sync"

	"github.com/vmihailenco/msgpack/v5"
)

// Subscriber represents one active WATCH subscription from a session.
type Subscriber struct {
	SessionID string
	StreamID  uint32
	Send      chan<- []byte // pre-encoded frame bytes; must be buffered
}

// FanoutBus manages per-key subscriber lists and broadcasts pre-encoded delta
// frames. The delta body is encoded once per mutation; per-subscriber cost is
// only the 4-element frame envelope with the subscriber's stream_id.
type FanoutBus struct {
	mu   sync.RWMutex
	subs map[string][]Subscriber
}

// NewFanoutBus creates an empty fanout bus.
func NewFanoutBus() *FanoutBus {
	return &FanoutBus{subs: make(map[string][]Subscriber)}
}

// Subscribe adds a subscriber for a state key.
func (fb *FanoutBus) Subscribe(key string, sub Subscriber) {
	fb.mu.Lock()
	defer fb.mu.Unlock()
	fb.subs[key] = append(fb.subs[key], sub)
}

// Unsubscribe removes a subscriber by session and stream ID.
func (fb *FanoutBus) Unsubscribe(key, sessionID string, streamID uint32) {
	fb.mu.Lock()
	defer fb.mu.Unlock()

	subs := fb.subs[key]
	for i, s := range subs {
		if s.SessionID == sessionID && s.StreamID == streamID {
			subs[i] = subs[len(subs)-1]
			fb.subs[key] = subs[:len(subs)-1]
			return
		}
	}
}

// RemoveSession purges all subscriptions for a disconnected session.
func (fb *FanoutBus) RemoveSession(sessionID string) {
	fb.mu.Lock()
	defer fb.mu.Unlock()

	for key, subs := range fb.subs {
		filtered := subs[:0]
		for _, s := range subs {
			if s.SessionID != sessionID {
				filtered = append(filtered, s)
			}
		}
		if len(filtered) == 0 {
			delete(fb.subs, key)
		} else {
			fb.subs[key] = filtered
		}
	}
}

// Broadcast encodes a delta body once, then composes a DELTA frame per
// subscriber with their specific stream_id. The sender session is skipped
// (the server writes its own DELTA directly on the hot path).
// Subscribers whose channel is full are skipped (slow-client protection).
func (fb *FanoutBus) Broadcast(delta DeltaEntry, senderSessionID string) {
	bodyBytes, err := msgpack.Marshal(delta)
	if err != nil {
		return
	}
	rawBody := msgpack.RawMessage(bodyBytes)

	fb.mu.RLock()
	subs := fb.subs[delta.Key]
	snapshot := make([]Subscriber, len(subs))
	copy(snapshot, subs)
	fb.mu.RUnlock()

	for _, sub := range snapshot {
		if sub.SessionID == senderSessionID {
			continue
		}
		frame, err := msgpack.Marshal([]any{
			uint8(FrameDelta),
			sub.StreamID,
			delta.Version,
			rawBody,
		})
		if err != nil {
			continue
		}
		select {
		case sub.Send <- frame:
		default:
		}
	}
}

// SubscriberCount returns the number of active watchers for a key.
func (fb *FanoutBus) SubscriberCount(key string) int {
	fb.mu.RLock()
	defer fb.mu.RUnlock()
	return len(fb.subs[key])
}
